### PR TITLE
FEATURE: Improvements for the Node Creation Dialog

### PR DESCRIPTION
**Pull request by danielkestler**
Add improvements to the Node Creation Dialog like autofocusing the filter input field, keyboard navigation and insertion, see #3031

**What I did**
- autofocus node type filtering input element
- auto select first matching node type
- add keyboard navigation for up/down navigating the node type selection
- add enter keypress for inserting node

**How I did it**
- extended the state to hold a list of all filtered node types and the user selection that can then be toggled via the keyboard up/down arrows 

**How to verify it**
- open node type insertion dialog
- enter some filter text into the filter input
- use up/down arrows to select node type
- press enter to insert node

Bildschirmfoto 2022-07-06 um 15 50 57

### CODE REVIEW

**Comment by danielkestler** [comment]
We should make the input field more prominent to better fit our changes.

Bildschirmfoto 2022-07-07 um 08 26 55

Bildschirmfoto 2022-07-07 um 08 29 11

**Comment by Sebobo** [comment]
Cool. Like it much better 👍

**Comment by jonnitto** [comment]
Do we really want the autofocus feature? If you edit the page on a device without a keyboard (e. g. iPad), the software keyboards pop up

**Comment by Sebobo** [comment]
I love having autofocus in general. 
Is it possible to have a feature/device check whether we enable that?

**Comment by jonnitto** [comment]
Unfortunately, currently, there is no direct way to detect if a virtual keyboard appears on the screen using JavaScript.

**Comment by markusguenther** [comment]
It is such a nice improvement. Do we want to discuss the autofocus this week? Or do we take it as it is?
We have still time until 17th of November (feature freeze)

@jonnitto

**Review by jonnitto — APPROVED** [review]

File: packages/neos-ui/src/Containers/Modals/SelectNodeType/nodeTypeFilter.js [hunk 1]
```diff
@@ -15,19 +15,31 @@ const NodeTypeFilter = ({onChange, filterSearchTerm, i18nRegistry}) => {
...
 
     return (
         <div className={style.nodeTypeDialogHeader__filter}>
-            {filterSearchTerm ? (
-                <IconButton icon="times" onClick={handleResetFilter}/>
-            ) : (
-                <Icon icon="filter" padded="right"/>
-            ) }
+            <div className={style.nodeTypeDialogHeader__filterIconSearch}>
+                <Icon icon="search" size="md"/>
```
Comment (markusguenther) [inline packages/neos-ui/src/Containers/Modals/SelectNodeType/nodeTypeFilter.js]: The size md is invalid. We have just `"lg","xs","sm","1x","2x","3x","4x","5x","6x","7x","8x","9x","10x"`

File: packages/neos-ui/src/Containers/Modals/SelectNodeType/index.js [hunk 2]
```diff
@@ -182,10 +188,97 @@ export default class SelectNodeType extends PureComponent {
...
+
+            if (filteredNodeTypes.length > 0) {
+                filteredNodeTypesFlat.push(...filteredNodeTypes);
+                filteredNodeTypesGrouped.push(value);
+            }
+
+            return true;
+        });
+
+        const focusedNodeType = (filterSearchTerm !== '' && filteredNodeTypesFlat.length > 0) ? filteredNodeTypesFlat[0].name : null;
```
Comment (markusguenther) [inline packages/neos-ui/src/Containers/Modals/SelectNodeType/index.js]: It happens that the property name is not defined. that leads to errors in the console.
```Uncaught TypeError: can't access property "name", filteredNodeTypesFlat[(index + 1)] is undefined```

File: packages/neos-ui/src/Containers/Modals/SelectNodeType/index.js [hunk 3]
```diff
@@ -182,10 +188,97 @@ export default class SelectNodeType extends PureComponent {
...
+    }
+
+    handleKeyDown = event => {
+        if (event.key === 'ArrowUp' || event.key === 'ArrowDown') {
+            event.preventDefault();
+
+            const {filteredNodeTypesFlat, focusedNodeType} = this.state;
+            const index = filteredNodeTypesFlat.findIndex(element => element.name === focusedNodeType);
+
+            if (event.key === 'ArrowUp') {
```
Comment (markusguenther) [inline packages/neos-ui/src/Containers/Modals/SelectNodeType/index.js:271]: Here we need to check if the index is valid ;)

File: packages/neos-ui/src/Containers/Modals/SelectNodeType/nodeTypeFilter.js (hunk 1 above)
Comment (markusguenther) [reply packages/neos-ui/src/Containers/Modals/SelectNodeType/nodeTypeFilter.js]: Replaced it now with `1x` if it is wrong, we can change it before the 8.2 release ;)

File: packages/neos-ui/src/Containers/Modals/SelectNodeType/index.js (hunk 2 above)
Comment (markusguenther) [reply packages/neos-ui/src/Containers/Modals/SelectNodeType/index.js]: that was not the root case but added a check to be save that it is working

File: packages/neos-ui/src/Containers/Modals/SelectNodeType/index.js (hunk 3 above)
Comment (markusguenther) [reply packages/neos-ui/src/Containers/Modals/SelectNodeType/index.js:271]: fixed that issue

**Review by markusguenther — APPROVED** [review]
Thanks, @danielkestler, for the improvements. Looking forward to more UX changes ;)

**Pull request merged**